### ISSUE TITLE
API not properly excluding Shadow DOM content

### ISSUE THREAD

**Issue by marcysutton** [dequelabs/axe-core#744]
I uncovered a bug when upgrading axe-matchers to axe-core@3x, where it would fail on Shadow DOM content that was supposed to be excluded through the API. I would expect that by excluding a shadow host element, it wouldn't return issues from nodes inside of it. I was able to reproduce it in axe-core, here's the basic fixture:

```html
<div id="container">
    <p id="test-passes">Passing text</p>
    <div id="shadow-host"></div>
</div>
<script>
    var host = document.getElementById('shadow-host');
    var shadowRoot = host.attachShadow({ mode: 'open' });
    shadowRoot.innerHTML =  '<div style="background-color: #333">' +
        '<p>Failing text</p>' +
        '<ul><p>Failing list</p></ul>' +
    '</div>';
</script>
```
When I call axe.run on it with `include: [['#container']], exclude: [['#shadow-host']]`, I get color contrast and list violations for the shadow content. It should exclude those, but it's including shadow content for both rules with a stated target of `#shadow-host`. 

Here's a failing test: https://github.com/dequelabs/axe-core/commit/0ef5bb980d648758ed8f2914b807a2bba33ad5e6

```javascript
(shadowSupported ? it : xit)
('should find no nodes in Shadow DOM', function (done) {
    var config = { runOnly: { type: 'rule', values: ['list', 'color-contrast'] } };
    axe.run({ include: [['#container']], exclude: [['#shadow-host']] }, config, function (err, results) {
        assert.isNull(err);
        assert.lengthOf(results.violations, 0, 'violations');
        assert.lengthOf(results.passes, 2, 'passes');
        done();
    });
});
```

**Comment by WilcoFiers** [comment]
Ugh... this one had me running around the axe-core source like it was a maze. Turns out the problem is that the Context was generating its own flattened trees, and because of that, the shadowIds weren't matching. I'll work on a fix.

**Comment by anutin544** [comment]
Ok
